torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -482,6 +488,43 @@ ProcessGroupNCCL::ProcessGroupNCCL(
     asyncErrorHandling_ = false;
   }
 
+  // Perform health check by initializing dummy communicators and destroying
+  // them. This will help indicate any NCCL-related issues prior to the first
+  // collective.
+  // Run it in a separate thread and wait on CV to handle timeouts, since
+  // majority of getNCCLComm failures are hangs.
+  std::mutex healthCheckMutex;
+  std::condition_variable healthCheckCv;
+  bool healthCheckSuccess = false;
+  auto t = std::thread([&healthCheckMutex, &healthCheckCv, &healthCheckSuccess, this]() {
+    std::vector<at::Device> rankDevice = {getDeviceForRank(rank_)};
+    const auto key = getKeyFromDevices(rankDevice);
+    // OpType does not matter, only need to set to not go through send/recv
+    // path.
+    auto& ncclComms = getNCCLComm(key, rankDevice, OpType::ALLREDUCE);
+    // Now destroy the communicators and remove them from cache so we don't use
+    // destroyed communicators.
+    destroyNCCLComms(key);
+    // Notify main thread the health check is complete.
+    {
+      std::lock_guard<std::mutex> lk(healthCheckMutex);
+      healthCheckSuccess = true;
+    }
+    healthCheckCv.notify_one();
+  });
+  // We don't need to join the thread, just need to verify health check via the
+  // CV. Hence we detach the thread here.
+  t.detach();
+  std::unique_lock<std::mutex> lock(healthCheckMutex);
+  LOG(INFO) << "Rank: " << rank_ << " will wait for: " << options->timeout.count() << " ms for NCCL health check to complete.";
```

Comment:
nit: This format is slightly different from the format followed in all other places: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L295

Reply:
Probably not a good idea to perform logging under a lock.

Reply:
Good catch, will remove it to before the lock.